trpcultivate_phenotypes/src/Plugin/Validators/ValidDataFile.php

```diff
@@ -0,0 +1,198 @@
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition){
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('entity_type.manager')
+    );
+  }
+
+  /**
+   * Perform validation of data file.
+   * Checks include:
+   *  - Parameter filename or file id is valid.
+   *  - Has Drupal File Id number assigned and can be loaded.
+   *  - Valid file extension and file mime type configured by the file importer instance.
+   *  - File exists and not empty.
+   *  - File can be opened.
+   *
+   * @param string $filename
+   *   The full path to a file within the file system (Absolute file path).
+   * @param integer $fid
+   *   [OPTIONAL] The unique identifier (fid) of a file that is managed by Drupal File System.
+   *
+   * @return array
+   *   An associative array with the following keys.
+   *    - case: a developer focused string describing the case checked.
+   *    - valid: either TRUE or FALSE depending on if the file is valid or not.
+   *    - failedItems: an array of "items" that failed to be used in the message to the user. This is an empty array if the file input was valid.
+   */
+  public function validateFile(string $filename, int|null $fid = NULL) {
+
+    // Parameter check, verify that the filename/file path is valid.
+    if (empty($filename) && is_null($fid)) {
+      return [
+        'case' => 'Filename is empty string',
+        'valid' => FALSE,
+        'failedItems' => ['filename' => '', 'fid' => $fid]
+      ];
+    }
+
+    // Parameter check, verify the file id number is not 0 or negative values.
+    if (!is_null($fid) && $fid <= 0) {
+      return [
+        'case' => 'Invalid file id number',
+        'valid' => FALSE,
+        'failedItems' => ['filename' => $filename, 'fid' => $fid]
+      ];
+    }
+
+    // Holds the file object when file is a managed file.
+    $file_object = NULL;
+
+    // Load file object.
+    if (is_numeric($fid) && $fid > 0) {
+      // The file input is integer value, the file id number. Load the file object by fid number.
+      $file_object = File::load($fid);
+    }
+    elseif ($filename) {
+      // The file input is a string value, a path to the file. Locate the file entity
+      // by uri and load the file object using the returned file id number that matched.
+      $file_entities = $this->service_EntityTypeManager
+        ->getStorage('file')
+        ->loadByProperties(['uri' => $filename]);
+
+      $file_entity = reset($file_entities);
+
+      if ($file_entity) {
+        $fid = $file_entity->get('fid')->value;
+        $file_object = File::load($fid);
+      }
+    }
+
+    // Check that the file input provided returned a file object.
+    if (is_null($file_object)) {
+      return [
+        'case' => 'Filename or file id failed to load a file object',
```

Comment:
It would be good to know if the filename or the fid were used here :thinking. I think my pseudocode showed this orientation but in practice I now think there should be a check for the file object right after it's loaded above (so two checks: one after loaded via fid and the other after loaded via filename).